src/core/Lockfile_target.ml

```diff
@@ -1,25 +1,76 @@
-   Should contain a list of all dependencies used by
-   a project, at exact versions. Can be attached to a code target,
-   or a target by itself.
-*)
 type t = {
-  lockfile : Fpath.t;
-  lockfile_kind : Lockfile_kind.t;
-  lazy_lockfile_content : string lazy_t;
-  lazy_lockfile_ast_and_errors : Dependency.t list lazy_t;
-  manifest_target : manifest_target option;
+  source : Source.t;
+      (** The source of the data as is relevant to the user. This could be, e.g., a
+      relative (from the project root) path to a file, a git object and
+      associated information, or anything else a Source.t can designate.
+
+      This should be used when reporting a location to the user. *)
+  internal_path_to_content : Fpath.t;
+      (** The path to a file which contains the data to be scanned. This could be
+      the same as the source, if the source is a path to a regular file (or an
+      absolute path to the same), or it could be a tempfile. This should be
+      used to obtain the contents of the target, but not for reporting to the
+      user, other than possibly for debugging purposes. *)
+  kind : Lockfile_kind.t;
+      (** The type of lockfile this is. Analogous to analyzer for a source code
+        target. *)
+  manifest : manifest_target option;
+      (** Optionally, a manifest file associated with this lockfile. *)
+  lazy_content : string lazy_t;
+      (** The contents of the lockfile, as a string. *)
+  lazy_ast_and_errors : Dependency.t list lazy_t;
+      (** The parsed contents of the lockfile, comprising the list of specified
+          dependencies and their versions. *)
 }
+(** A lockfile to be scanned by a supply chain rule.
+
+   A lockfile is a file which specifies exact versions of every dependency,
+   including transitive dependencies, which belong to the "locked" package.
+   Typically this is generated from a more general {{!manifest_target}
+   "manifest" file}.
 
-(* A manifest file, which should contain a list of
-   direct dependencies (handwritten by a user).
-   Used in the parsing of some lockfiles, to determine
-   which dependencies are direct and which are transitive.
-   Can only ever be attached to a lockfile target.
+   Examples of lockfiles include [package-lock.json] (javascript/npm) or
+   [Cargo.lock] (rust/cargo).
+
+   Can be attached to a {{!Xtarget.t} code target}, or it can be a standalone
+   target by itself.
 *)
+
 and manifest_target = {
-  manifest : Fpath.t;
-  manifest_kind : Manifest_kind.t;
-  lazy_manifest_content : string lazy_t;
-  lazy_manifest_ast_and_errors : Dependency.manifest_dependency list lazy_t;
+  source : Source.t;
+      (** The source of the data as is relevant to the user. This could be, e.g., a
+      relative (from the project root) path to a file, a git object and
+      associated information, or anything else a Source.t can designate.
+
+      This should be used when reporting a location to the user. *)
+  internal_path_to_content : Fpath.t;
```

Comment:
the fact that both time we have this source and internal_path_to_content is a bad smell; probably there is a better abstraction to design.

Reply:
Based on discussion with Martin and other comments here, I have reworked the organization of many of the additional/changed types in fc35371b4.